### PR TITLE
encryption: support encryption on top of custom reader and writer

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What problem does this PR solve?

Problem Summary:

Unable to use existing encryption facility in [raft engine](https://github.com/tikv/raft-engine).

### What is changed and how it works?

What's Changed:

- Allow creating encrypted file on top of arbitrary writable/readable objects
- Support random read for encryption reader
- Reduce buffer allocation of encryption reader/writer
- Fix `EncrypterWriter` bugs when using Plaintext encryption method or underlying writer reports EOF

### Check List 

Tests
- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by tabokie** [comment]
@yiwu-arbug I'm not familiar with encryption algorithm to know the proper placement for things like `Iv::seek`, `AES_BLOCK_SIZE`, and where to reject settings that don't support random read. PTAL~

**Review by yiwu-arbug — COMMENTED** [review]
rest LGTM

File: components/encryption/src/io.rs [hunk 1]
```diff
@@ -138,21 +196,31 @@ impl<R> CrypterReader<R> {
...
         }
         self.do_crypter(buf, read_count)
     }
 }
 
+impl<R: Seek> Seek for CrypterReader<R> {
+    fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
+        let offset = self.reader.seek(pos)?;
+        if self.method != EncryptionMethod::Plaintext {
+            self.init_crypter(offset)?;
```
[the remark this answers is not in the document]
Comment (tabokie) [reply components/encryption/src/io.rs:218]: I think here each seek will call `init_crypter` to reset IV

File: components/encryption/src/io.rs [hunk 2]
```diff
@@ -108,27 +129,64 @@ impl<R> CrypterReader<R> {
...
+                crypter_buffer: Vec::new(),
             },
             iv,
         ))
     }
 
+    fn reserve_buffer(&mut self, size: usize) {
+        // OCrypter require the output buffer to have block_size extra bytes, or it will panic.
+        if size + self.block_size > self.crypter_buffer.len() {
+            self.crypter_buffer.resize(size + self.block_size, 0);
```
[the remark this answers is not in the document]
Comment (tabokie) [reply components/encryption/src/io.rs:150]: I think for `CrypterWriter`, which is append-only and un-reusable, keeping a large buffer is fine. For `CrypterReader`, I'll set an upper bound and do a streaming style encryption if read request is too large.

File: components/encryption/src/io.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply components/encryption/src/io.rs]: sorry, I misread the logic.

**Review by yiwu-arbug — APPROVED** [review]

**Review by hunterlxt — APPROVED** [review]
Roughly reviewed LGTM

File: components/encryption/src/io.rs [hunk 3]
```diff
@@ -16,6 +16,9 @@ use crate::{Iv, Result};
 use file_system::File;
 use tikv_util::box_err;
 
+const AES_BLOCK_SIZE: usize = 16;
+const MAX_CRYPTER_READER_BUFFER_SIZE: usize = 1024;
```
Comment (yiwu-arbug) [inline components/encryption/src/io.rs:20]: 1K is too small, maybe 1M is better.

**Pull request merged**